File: doc-examples/azure-cosmos-example-kotlin/src/main/kotlin/example/ItemPriceAttributeConverter.kt
package example

import io.micronaut.core.convert.ConversionContext
import io.micronaut.data.model.runtime.convert.AttributeConverter
import jakarta.inject.Singleton

@Singleton // <1>
class ItemPriceAttributeConverter : AttributeConverter<ItemPrice?, Double?> {

    // <2>
    override fun convertToPersistedValue(itemPrice: ItemPrice?, context: ConversionContext): Double? {
        return itemPrice?.price
    }

    // <3>
    override fun convertToEntityValue(value: Double?, context: ConversionContext): ItemPrice? {
        return if (value == null) null else ItemPrice(value)
    }

}

File: doc-examples/azure-cosmos-example-python/src/main/python/example/ItemPriceAttributeConverter.py
from jakarta.inject import Singleton
from micronaut.core.convert import ConversionContext
from micronaut.data.model.runtime.convert import AttributeConverter

from example.ItemPrice import ItemPrice


@Singleton
class ItemPriceAttributeConverter(AttributeConverter[ItemPrice, float]):

    def convertToPersistedValue(self, book_price: ItemPrice | None, context: ConversionContext) -> float | None:
        return None if book_price is None else book_price.price

    def convertToEntityValue(self, value: float | None, context: ConversionContext) -> ItemPrice | None:
        return None if value is None else ItemPrice(value)
